lib/services/message-service.ts

```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    /**
+     * Instruct the child to start listening if we haven't already
+     */
+    public listen() {
+        if(!this.listening) {
+            this.listening = true;
+            this._startListening();
+        }
+    }
+
+    /**
+     * Express an interest in a particular type of event
+     * @param registration Object detailing event type, callback, etc
+     */
+    public registerEvent(registration: ServiceRegistration): void {
+        // For each event type being registered
+        for(const event of registration.events) {
+            // Ensure we have a listener array for it
+            if(this._eventListeners[event] == null) {
+                this._eventListeners[event] = [];
+            }
+            // Store the expression of interest
+            this._eventListeners[event].push(registration);
+        }
+    }
+
+    /**
+     * Emit data to the external service
+     * @param data ServiceEmitRequest to parse
+     */
+    public sendData(data: ServiceEmitRequest): Promise<MessageEmitResponse> {
+        // Check the contexts for relevance before passing down the inheritance
+        if(data.contexts[this.serviceName]) {
+            return this._sendData(data.contexts[this.serviceName]);
+        } else {
+            // If we have a context to emit to this service, then no-op is correct resolution
+            return Promise.resolve({
+                err: new Error(`No ${this.serviceName} context`),
+                source: this.serviceName,
+            });
+        }
+    }
+
+    /**
+     * Enqueue a MessageWorkerEvent
+     * @param data event to enqueue
+     */
+    public queueEvent(data: MessageWorkerEvent) {
+        // This simply passes it up the chain...
+        // but ensures that MessageServices only enqueue the correct type
+        super.queueEvent(data);
+    }
+
+    // TODO: event should be in the same format as emitted
+    /**
+     * Retrieve the comments in a thread that match an optional filter
+     * @param event details to identify the event
+     * @param filter regex of comments to match
+     */
+    public fetchThread(_event: ReceiptContext, _filter: RegExp): Promise<string[]> {
+        return Promise.reject(new Error('Not yet implemented'));
+    }
+
+    // TODO: event should be in the same format as emitted
+    /**
+     * Retrieve the private message history with a user
+     * @param event details of the event to consider
+     * @param filter optional criteria that must be met
+     */
+    public fetchPrivateMessages(_event: ReceiptContext, _filter: RegExp): Promise<string[]> {
+        return Promise.reject(new Error('Not yet implemented'));
+    }
+
+    /**
+     * Activate this object as a listener
+     */
+    protected abstract _startListening(): void;
+
+    /**
+     * Emit data to the API
+     * @param data emit context
+     */
+    protected abstract _sendData(data: ServiceEmitContext): Promise<MessageEmitResponse>
```

Comment:
If we're not actually letting client bots include any of these derived services directly, then we shouldn't be using names such as `_sendData`, as this confuses what's actually happening under the hood, esp. as that's the normal public method.